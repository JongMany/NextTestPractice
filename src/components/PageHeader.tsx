"use client";
import { usePathname } from "next/navigation";
import React from "react";
import styled from "styled-components";
import GoBack from "./styled/GoBack";

const Container = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #1e40ff;
`;
const Title = styled.h1`
  padding: 20px;
  color: #ffffff;
  font-size: 20px;
  font-weight: 700;
`;

export default function PageHeader() {
  const pathname = usePathname();
  let title = "";

  if (pathname === "/add") title = "할 일 추가";
  else if (pathname === "/") title = "할 일 목록";
  else if (pathname.startsWith("/detail")) title = "할 일 상세";
  else title = "에러😕";

  return (
    <Container>
      <Title>{title}</Title>
      {pathname !== "/" && <GoBack href="/">돌아가기</GoBack>}
    </Container>
  );
}
