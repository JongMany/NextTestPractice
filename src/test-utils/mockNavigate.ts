// jest.mock("next/navigation", () => ({
//   ...require("next-router-mock"),
//   useSearchParams: jest.fn(),
// }));

// export default 