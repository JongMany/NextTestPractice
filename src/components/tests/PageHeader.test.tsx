// https://github.com/vercel/next.js/discussions/42527

import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";
import PageHeader from "@/components/PageHeader";
import { mockRouter } from "@/test-utils/next-router-utils";
import renderer from "react-test-renderer";

// https://github.com/amannn/next-intl/discussions/331
const mockUsePathname = jest.fn();

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  usePathname() {
    return mockUsePathname();
  },
}));

describe("<PageHeader/>", () => {
  it("renders component correctly", () => {
    mockUsePathname.mockImplementation(() => "/");

    render(<PageHeader />);
    const label = screen.getByText("할 일 목록");
    expect(label).toBeInTheDocument();
    const goBack = screen.queryByText("돌아가기");
    expect(goBack).not.toBeInTheDocument();
  });

  it("renders component correctly snapshot", () => {
    mockUsePathname.mockImplementation(() => "/");
    const tree = renderer.create(<PageHeader />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders component correctly when pathname is "/add"', () => {
    mockUsePathname.mockImplementation(() => "/add");
    render(<PageHeader />);

    const label = screen.getByText("할 일 추가");
    expect(label).toBeInTheDocument();
    const goBack = screen.queryByText("돌아가기");
    expect(goBack).toBeInTheDocument();
    expect(goBack).toHaveAttribute("href", "/");
  });

  it("renders component correctly snapshot when pathname is '/add'", () => {
    mockUsePathname.mockImplementation(() => "/add");
    const tree = renderer.create(<PageHeader />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders component correctly when pathname is "/detail/:id"', () => {
    mockUsePathname.mockImplementation(() => "/detail/1");

    render(<PageHeader />);
    const label = screen.getByText("할 일 상세");
    expect(label).toBeInTheDocument();
    const goBack = screen.queryByText("돌아가기");
    expect(goBack).toBeInTheDocument();
    expect(goBack).toHaveAttribute("href", "/");
  });

  it("renders component correctly with Error", () => {
    mockUsePathname.mockImplementation(() => "/error");

    render(<PageHeader />);
    const label = screen.getByText("에러😕");
    expect(label).toBeInTheDocument();
    const goBack = screen.queryByText("돌아가기");
    expect(goBack).toBeInTheDocument();
    expect(goBack).toHaveAttribute("href", "/");
  });

  it("renders component correctly with goBack Link", async () => {
    mockUsePathname.mockImplementation(() => "/error");

    render(<PageHeader />, { wrapper: MemoryRouterProvider });

    const goBack = screen.getByText("돌아가기");
    fireEvent.click(goBack);

    // https://www.npmjs.com/package/next-router-mock#example-nextlink-with-react-testing-library
    // 이동한 경로가 "/"인지 확인
    // console.log(mockRouter);
    await waitFor(() => {
      expect(mockRouter.pathname).toEqual("/");
    });
  });
});
