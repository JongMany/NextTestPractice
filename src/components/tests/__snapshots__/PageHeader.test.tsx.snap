// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageHeader/> renders component correctly snapshot 1`] = `
<div
  className="sc-fa2a6099-0 fIvrUc"
>
  <h1
    className="sc-fa2a6099-1 jecMvq"
  >
    할 일 목록
  </h1>
</div>
`;

exports[`<PageHeader/> renders component correctly snapshot when pathname is '/add' 1`] = `
<div
  className="sc-fa2a6099-0 fIvrUc"
>
  <h1
    className="sc-fa2a6099-1 jecMvq"
  >
    할 일 추가
  </h1>
  <a
    className="sc-70ce8365-0 jqGvbP"
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    돌아가기
  </a>
</div>
`;
